Feat(eos_cli_config_gen): Add support for maintenance groups

c-po opened the pull request:
## Change Summary

<!-- Enter short PR description -->

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [x] Documentation content changes
- [ ] Other (please describe):

## Related Issue(s)

Fixes #1014

## Component(s) name

`arista.avd.eos_cli_config_gen`

## Proposed changes
<!--- Describe your changes in detail -->
<!--- Describe data model implemented for new features -->

```yaml
group:
  bgp:
    < group_name >:
      vrf: "< vrf_name >"
      neighbors:
        - "< ip_address >"
        - "< ipv6_address >"
        - "< peer_group_name >"
      maintenance_profile:
        bgp:
          - < profile_name >
  interface:
    < group_name >:
      interfaces:
        - "< interface_or_interface_range >"
      maintenance_profile:
        bgp:
          - < profile_name >
        interface:
          - < profile_name >
```

## How to test
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code has been rebased from devel before I start
- [x] I have read the [**CONTRIBUTING**](https://avd.sh/en/latest/docs/contributing/) document.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have updated [molecule CI](https://github.com/aristanetworks/ansible-avd/tree/devel/ansible_collections/arista/avd/molecule) testing accordingly
- [x] All new and existing tests passed ([`pre-commit`](https://www.avd.sh/docs/installation/development/#python-virtual-environment), `make linting` and `make sanity-lint`).

ClausHolbechArista commented on ansible_collections/arista/avd/molecule/eos_cli_config_gen/inventory/host_vars/group.yml:
```diff
@@ -0,0 +1,31 @@
+      vrf: red
+
+  interface:
+    SFP+_Interface_Group:
```
The `+` is not accepted by EOS. Please use valid data for tests.

ClausHolbechArista commented on ansible_collections/arista/avd/molecule/eos_cli_config_gen/inventory/host_vars/group.yml:
```diff
@@ -0,0 +1,31 @@
+
+    QSFP_Interface_Group:
+      interfaces:
+        - Ethernet11/-28/1
```
Interface range is not valid

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/README.md:
```diff
@@ -37,6 +37,9 @@
       - [Peer Filters](#peer-filters)
       - [Route Maps](#route-maps)
+    - [Groups](#groups)
+      - [BGP](#group-bgp)
```
heading should match TOC.
Group BGP
Group Interface

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/README.md:
```diff
@@ -546,6 +549,36 @@ route_maps:
 
+### Groups
+
+#### Group Interfaces
```
Use singular to match the data model

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/README.md:
````diff
@@ -546,6 +549,36 @@ route_maps:
+        interface: < interface_maintenance_profile_name >
+```
+
+#### BGP Interfaces
````
Wrong heading. (Also reorder to BGP comes first)

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/group.j2:
```diff
@@ -0,0 +1,36 @@
+{% if group.bgp is arista.avd.defined %}
```
This test is covered by `arista.avd.natural_sort` so it can be removed.

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/README.md:
```diff
@@ -546,6 +549,36 @@ route_maps:
+  bgp:
+    < group_name >:
+      vrf: "< vrf_name >"
+      neighbor:
```
Use plural in data model (sorry if I missed this on your issue)

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/README.md:
```diff
@@ -546,6 +549,36 @@ route_maps:
+        - "< ipv6_address >"
+        - "< peer_group_name >"
+      maintenance_profile:
+        bgp: < bgp_maintenance_profile_name >
```
It is possible to configure more than one maintenance profile bgp per group (same for interface):
```eos
group interface test
   interface Et10-20,30-40
   maintenance profile bgp test2
   maintenance profile bgp test
   exit
```
So we should make these into lists.

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/group.j2:
```diff
@@ -0,0 +1,36 @@
+{%     endif %}
+{%   endfor %}
+{% endif %}
+{% if group.interface is arista.avd.defined %}
```
This test is covered by arista.avd.natural_sort so it can be removed.

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/group.j2:
```diff
@@ -0,0 +1,36 @@
+!
+group interface {{ group_name }}
+{%     if group.interface[group_name].interfaces is arista.avd.defined %}
+{%       for interface in group.interface[group_name].interfaces %}
```
you can consider using `arista.avd.natural_sort` to simplify the template, since it covers the test above.

ClausHolbechArista reviewed (dismissed):
Some changes to the data model after all (sorry about that). Also you are missing documentation for this feature. I think it would make sense to have "group" and "maintenance" under a common heading like "maintenance mode".

Commit:
eos_cli_config_gen(feature): add grouping support for maintenance mode (#1014)

EOS supports grouing interfaces or BGP neighbors and assign them an individual
maintenance profile which is executed once the device enters maintenance mode.

This commit extends the eos_cli_config_gen datamodel an adds support for
grouping BGP neighbors and interfaces using the following data model:

group:
  bgp:
    < group_name >:
      vrf: "< vrf_name >"
      neighbors:
        - "< ip_address >"
        - "< ipv6_address >"
        - "< peer_group_name >"
      maintenance_profile:
        bgp:
          - < profile_name >
  interface:
    < group_name >:
      interfaces:
        - "< interface_or_interface_range >"
      maintenance_profile:
        bgp:
          - < profile_name >
        interface:
          - < profile_name >

Commit:
Upload artefact for issue #1014

c-po commented:
Adjusted to requested changes.

> I think it would make sense to have "group" and "maintenance" under a common heading like "maintenance mode".

Well it can be adjusted to:
```yml
maintenance_mode:
  group:
    bgp:
      < group_name >:
        vrf: "< vrf_name >"
        neighbors:
          - "< ip_address >"
          - "< ipv6_address >"
          - "< peer_group_name >"
        maintenance_profile:
          bgp:
            - < profile_name >
   interface:
      < group_name >:
        interfaces:
          - "< interface_or_interface_range >"
        maintenance_profile:
          bgp:
            - < profile_name >
          interface:
            - < profile_name >
```

I have no personal preference here and are open to whatever you prefer.

ClausHolbechArista commented:
> Adjusted to requested changes.
> 
> > I think it would make sense to have "group" and "maintenance" under a common heading like "maintenance mode".
> 
> Well it can be adjusted to:
> 
> ```yaml
> maintenance_mode:
>   group:
>     bgp:
>       < group_name >:
>         vrf: "< vrf_name >"
>         neighbors:
>           - "< ip_address >"
>           - "< ipv6_address >"
>           - "< peer_group_name >"
>         maintenance_profile:
>           bgp:
>             - < profile_name >
>    interface:
>       < group_name >:
>         interfaces:
>           - "< interface_or_interface_range >"
>         maintenance_profile:
>           bgp:
>             - < profile_name >
>           interface:
>             - < profile_name >
> ```
> 
> I have no personal preference here and are open to whatever you prefer.

Sorry I was not clear. I meant only in the documentation, to have the two sections under a common main heading.

c-po commented:
Something like the following in README.md?

### Maintenance
#### Group

```yaml
group:
  bgp:
    < group_name >:
      vrf: "< vrf_name >"
      neighbors:
        - "< ip_address >"
        - "< ipv6_address >"
        - "< peer_group_name >"
      maintenance_profile:
        bgp:
          - < profile_name >
  interface:
    < group_name >:
      interfaces:
        - "< interface_or_interface_range >"
      maintenance_profile:
        bgp:
          - < profile_name >
        interface:
          - < profile_name >
```

ClausHolbechArista commented:
> Something like the following in README.md?

Once again I managed to be unclear. This PR has no template for `eos_cli_config_gen` auto generated device documentation. Before you spend time on it, we should sort out the two "competing" PRs covering this. (See comment on issue)

c-po commented:
Ah, now I got the point. Yes I did not spend any time on that documentation templating. I Guess you are referring to PR #1020 which also has a implementation for this topic with a slightly different datamodel. I also see a bunch of conflicting files.

It would be unhonest to not say "please use my PR" but I get your point - we should select the one which is best for all worlds. I still recommend to maybe split this in two PRs so we both can work on the maintenance implementation. Still happy to add some initial work of the documentation templating which can then later be referenced in the other PR.

Another solution would be to take this PR "as is" and then use the documentation part from #1020 to polish this one up.

c-po commented:
Somehow GitHub thinks there are still changes requested - or am I missing something?

ClausHolbechArista commented:
No this is just waiting for us to decide on the data model. I hope to find out by tomorrow.

ClausHolbechArista commented:
@c-po The maintainer group prefers the data model in this PR, so when you add the device documentation templates I can do a final review and merge it. Thank you for the patience :)

daanvdsanden commented:
I also left a comment in #1020, but I feel this model is not consistent with the rest of the models used, see for reasoning and examples my comments in #1020. I can live with both implementations, since they both work and give a working configuration.

c-po committed:
eos_cli_config_gen(feature): add maintenance group documentation template (#1014)

c-po committed:
Upload artefact for issue #1014

c-po commented:
@ClausHolbechArista I hope this suits your expectation.

ClausHolbechArista commented:
> @ClausHolbechArista I hope this suits your expectation.

Not entirely. We should not generate documentation if the feature is not set. Is it okay that I adjust the templates directly on your branch?

c-po commented:
Of course @ClausHolbechArista just go ahead. Thanks

ClausHolbechArista committed:
Rework device documentation and adjust data models

ClausHolbechArista committed:
Update readme

titom73 reviewed (approved):
LGTM!

carlbuchmann commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos-intended-config.j2:
```diff
@@ -126,6 +126,10 @@
 {% include 'eos/virtual-source-nat.j2' %}
 {# event handler #}
 {% include 'eos/event-handler.j2' %}
+{# BGP Groups configuration #}
```
This is more cosmetic, but the ordered generate doesn't match what I tested on 4.25.4M DUT.

```shell
vrf instance MGMT
!
group bgp bar
   vrf red
   neighbor peer-group-baz
   maintenance profile bgp downlink-neighbors
   exit
!
group bgp foo
   neighbor fe80::1
   neighbor 169.254.1.1
   maintenance profile bgp uplink-neighbors
   maintenance profile bgp ixp
   exit
!
interface Port-Channel5
!
interface Ethernet1
!
interface Loopback0
!
interface Loopback1
!
interface Management1
   description oob_management
   vrf MGMT
   ip address 192.168.200.110/24
!
interface Vlan4093
   description MLAG_PEER_L3_PEERING
   ip address 10.255.251.6/31
!
interface Vlan4094
   description MLAG_PEER
   no autostate
   ip address 10.255.252.6/31
!
interface Vxlan1
   vxlan source-interface Loopback1
   vxlan virtual-router encapsulation mac-address mlag-system-id
   vxlan udp-port 4789
!
group interface QSFP_Interface_Group
   interface Et1,5
   maintenance profile interface uplink-interfaces
   exit
!
group interface SFP_Interface_Group
   maintenance profile interface ix-interfaces
   maintenance profile interface downlink-interfaces
   maintenance profile bgp local-ix
   maintenance profile bgp downlink-neighbors
   exit
!
```

carlbuchmann reviewed (approved):
LGTM!
I noted a cosmetic suggestion to adjust to CLI output order, but we can update this in another PR - I will merge!

titom73 reviewed (approved):
LGTM!
